File: java/src/array/DoublyLinkedList.java
package array;

public class DoublyLinkedList<E> {
    private Node head;
    private Node tail;
    private int size;

    private class Node {
        private E data;
        private Node next;
        private Node prev;

        public Node(E data) {
            this.data = data;
            this.next = null;
            this.prev = null;
        }

        public String toString() {
            return String.valueOf(this.data);
        }
    }

    public String toString() {
        Node tmp = head;

        StringBuffer sb = new StringBuffer();
        sb.append("[");
        for(int i = 0 ; i < size ; i++) {
            sb.append(tmp.data);
            sb.append(",");
            tmp = tmp.next;
        }
        sb.deleteCharAt(sb.length() - 1);
        sb.append("]");

        return sb.toString();
    }

    public void addFirst(E input) {
        Node node = new Node(input);
        node.next = head;

        if (head != null)
            head.prev = node;

        head = node;

        if (tail == null)
            tail = node;

        size++;
    }

    public void addLast(E input) {
        if (head == null) {
            addFirst(input);
        } else {
            Node node = new Node(input);
            node.prev = tail;
            tail.next = node;
            tail = node;

            size++;
        }
    }

    private Node node(int idx) {
        if (idx < size / 2) {
            Node tmp = head;

            for(int i = 0; i < idx; i++)
                tmp = tmp.next;

            return tmp;
        } else {
            Node tmp = tail;

            for(int i = size - 1; i > idx; i--)
                tmp = tmp.prev;

            return tmp;
        }
    }

    public void add(int idx, E input) {
        if (idx == 0) {
            addFirst(input);
        } else if (size == idx) {
            addLast(input);
        } else {
            Node next = node(idx);
            Node prev = next.prev;
            Node newNode = new Node(input);

            prev.next = newNode;
            newNode.prev = prev;
            newNode.next = next;
            next.prev = newNode;

            size++;
        }
    }

    public E removeFirst() {
        if (head == null)
            return null;

        Node delete = head;
        head = delete.next;

        if (head != null) {
            head.prev = null;
        }

        E returnValue = delete.data;
        delete = null;
        size--;

        return returnValue;
    }

    public E remove(int idx) {
        if (idx == 0) {
            return removeFirst();
        } else if (idx < size) {
            Node delete = node(idx);

            Node prev = delete.prev;
            Node next = delete.next;
            prev.next = next;
            if (next != null)
                next.prev = prev;
            else
                tail = prev;

            E returnValue = delete.data;
            delete = null;
            size--;

            return returnValue;
        }

        return null;
    }

    public E removeLast() {
        if (size == 0)
            return null;
        return remove(size - 1);
    }

    public int size() {
        return this.size;
    }

    public E get(int idx) {
        if (head == null || idx >= size)
            return null;

        return node(idx).data;
    }

    public int indexOf(E data) {
        Node tmp = head;

        for (int i = 0 ; i < size ; i++) {
            if (tmp.data.equals(data))
                return i;
            tmp = tmp.next;
        }

        return -1;
    }

    public ListIterator listIterator() {
        return new ListIterator();
    }

    class ListIterator {
        private Node lastReturned;
        private Node next;
        private int nextIdx;

        ListIterator() {
            next = head;
        }

        public E next() {
            lastReturned = next;
            next = next.next;
            nextIdx++;

            return lastReturned.data;
        }

        public boolean hasNext() {
            return nextIdx < size();
        }

        public E previous() {
            lastReturned = next = (next == null) ? tail : next.prev;

            nextIdx--;
            return lastReturned.data;
        }

        public boolean hasPrevious() {
            return nextIdx > 0;
        }

        public void add (E input) {
            Node node = new Node(input);
            node.next = next;

            if (lastReturned == null) {
                head = node;
            } else {
                node.prev = lastReturned;
                lastReturned.next = node;
            }

            if (next == null)
                tail = node;
            else
                next.prev = node;

            lastReturned = node;
            next = node.next;
            size++;
            nextIdx++;
        }

        public void remove() {
            if (lastReturned == null)
                throw new IllegalStateException();

            Node prevDel = lastReturned.prev;
            Node del = lastReturned;
            Node nextDel = lastReturned.next;

            if (prevDel == null)
                head = nextDel;
            else
                prevDel.next = nextDel;

            if (nextDel == null)
                tail = prevDel;
            else
                nextDel.prev = prevDel;

            del = null;
            lastReturned = prevDel == null ? head : prevDel;
            next = lastReturned.next;

            size--;
            nextIdx--;
        }
    }
}

File: python/array/doublyLinkedList.py
class DoublyLinkedList:
    head = None
    tail = None
    length = 0

    class Node:
        next = None
        prev = None
        def __init__(self, data):
            self.data = data
    
    def toString(self):
        tmp = self.head
        result = []
        for i in range(self.length):
            result.append(tmp.data)
            tmp = tmp.next
        return ','.join(result)

    def addFirst(self, input):
        node = self.Node(input)
        node.next = self.head

        if self.head is not None:
            self.head.prev = node

        if self.tail is None:
            self.tail = node

        self.head = node
        self.length += 1

    def addLast(self, input):
        node = self.Node(input)
        node.prev = self.tail

        if self.tail is not None:
            self.tail.next = node

        if self.head is None:
            self.head = node

        self.tail = node
        self.length += 1

    def __node(self, idx):
        if idx <= self.length // 2:
           tmp = self.head
           for i in range(idx):
               tmp = tmp.next
        else:
            tmp = self.tail
            for i in range(self.length - 1, idx, -1):
                tmp = tmp.prev

        return tmp

    def addFront(self, idx, input):
        if idx == 0:
            self.addFirst(input)
        elif idx >= self.length:
            self.addLast(input)
        else:
            node = self.Node(input)
            nextNode = self.__node(idx)

            node.prev = nextNode.prev
            nextNode.prev.next = node
            node.next = nextNode
            nextNode.prev = node

            self.length += 1

    def addBack(self, idx, input):
        if idx >= self.length:
            self.addLast(input)
        else:
            node = self.Node(input)
            prevNode = self.__node(idx)
            
            if prevNode is None:
                self.head = node
            else:
                node.next = prevNode.next
                node.prev = prevNode
                prevNode.next = node

            if node.next is None:
                self.tail = node
            else:
                node.next.prev = node

            self.length += 1

    def removeFront(self):
        if self.head is None:
            return None

        deleteNode = self.head
        self.head = deleteNode.next
        result = deleteNode.data
        deleteNode = None
        self.length -= 1

        return result

    def removeBack(self):
        if self.tail is None:
            return None

        deleteNode = self.tail
        self.tail = deleteNode.prev
        result = deleteNode.data
        deleteNode = None
        self.length -= 1

        return result

    def remove(self, idx):
        if idx >= self.length:
            return None
        
        deleteNode = self.__node(idx)
        prevNode = deleteNode.prev
        nextNode = deleteNode.next

        if prevNode is None:
            self.head = nextNode
        else:
            prevNode.next = nextNode

        if nextNode is None:
            self.tail = prevNode
        else:
            nextNode.prev = prevNode

        result = deleteNode.data
        deleteNode = None
        self.length -= 1

        return result

    def size(self):
        return self.length

    def empty(self):
        return self.head is None

    def get(self, idx):
        if idx >= self.length:
            return None
        return self.__node(idx).data

    def indexOf(self, data):
        tmp = self.head
        for i in range(self.length):
            if data == tmp.data:
                return i
            tmp = tmp.next
        return -1



